source/common/buffer/zero_copy_input_stream_impl.cc

```diff
@@ -0,0 +1,54 @@
+#include "common/buffer/zero_copy_input_stream_impl.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Buffer {
+
+ZeroCopyInputStreamImpl::ZeroCopyInputStreamImpl(Buffer::Instance& buffer) {
+  move(buffer);
+  finish();
+}
+
+void ZeroCopyInputStreamImpl::move(Buffer::Instance& instance) {
+  ASSERT(!finished_);
+
+  buffer_.move(instance);
+}
+
+bool ZeroCopyInputStreamImpl::Next(const void** data, int* size) {
+  if (position_ != 0) {
+    buffer_.drain(position_);
+    position_ = 0;
+  }
+
+  Buffer::RawSlice slice;
+  const uint64_t num_slices = buffer_.getRawSlices(&slice, 1);
+
+  if (num_slices > 0) {
+    *data = slice.mem_;
+    *size = slice.len_;
+    position_ = slice.len_;
+    byte_count_ += slice.len_;
+    return true;
+  }
+
+  if (!finished_) {
```

Comment:
It is callers responsibility to maintain. Added comment. transcoding interface add BytesAvailable() to avoid this.